File: Diagrammatize.Design.Pattern/code/cxx/State/A3/NoonState.cpp
#include "NoonState.hpp"
#include "DayState.hpp"
#include "NightState.hpp"
#include "Context.hpp"

NoonState::NoonState() {
}

State* NoonState::getInstance() {
    static NoonState singleton;
    return &singleton;
}

void NoonState::doClock(Context* context, int hour) {
    if (hour < 9 || 17 <= hour) {
        context->changeState(NightState::getInstance());
    } else if (9 <= hour && hour < 12 || 13 <= hour && hour < 17) {
        context->changeState(DayState::getInstance());
    }
}

void NoonState::doUse(Context* context) {
    context->callSecurityCenter("紧急：午餐时间使用金库！");
}

void NoonState::doAlarm(Context* context) {
    context->callSecurityCenter("按下警铃(午餐时间)");
}

void NoonState::doPhone(Context* context) {
    context->recordLog("午餐时间的通话录音");
}

std::string NoonState::toString() const {
    return "[午餐时间]";
}

File: Diagrammatize.Design.Patterns/code/python/State/A3/NoonState.py
from State import State

class NoonState(State):
    _singleton = None

    def __init__(self):
        print("create NoonState")
        pass

    @classmethod
    def getInstance(cls):
        if not cls._singleton:
            cls._singleton = cls()
        return cls._singleton

    def doClock(self, context, hour):
        from DayState import DayState
        from NightState import NightState
        if hour < 9 or 17 <= hour:
            context.changeState(NightState.getInstance())
        elif (9 <= hour < 12) or (13 <= hour < 17):
            context.changeState(DayState.getInstance())

    def doUse(self, context):
        context.callSecurityCenter("紧急：午餐时间使用金库！")

    def doAlarm(self, context):
        context.callSecurityCenter("按下警铃(午餐时间)")

    def doPhone(self, context):
        context.recordLog("午餐时间的通话录音")

    def __str__(self):
        return "[午餐时间]"


